### DISCUSSION TITLE
Wezterm Does Not Refresh the View to Previous Prompt Position  After Launching A Program

### DISCUSSION THREAD

**Discussion by sidzwans** [wezterm/wezterm#3193] [General]
My distro is Arch linux, using Sway (Wayland) WM. I have been a long time user of Termite until recently due to it no longer being maintained. During the short transition period to Alacritty (left because got problem with prompt of zsh theme did render in proportion to font size). Then to to Kitty (abandoned after left than 4 hours due to its not having VI mode and it won't be in the future by the way the maintainer reacted to feature request). So now I am using Wezterm and I've decided that I am going to stay for good.

I am very satisfied with wezterm except with one issue that I face. Whenever I close a program, the view doesn't refresh. 
So I was left with the previous last view of the program in the terminal. This is not the behavior of the terminals that I used before. Is this an expected behavior of wezterm, or is it a bug? Do I have to spawn the program to a new tab? It seems like an extra steps if I have to do that.

![wezterm_issue](https://user-images.githubusercontent.com/28080312/222922515-414fec66-d869-41f0-956c-f86e79c6da8c.png)

Can anyone suggest the workflow so I can open let's say Neovim in the same terminal, and the terminal will be as before when I close the program.

Thanks in advance. Regards.

**Comment by sidzwans** [comment]
Update: My shell is zsh. It seems that I encounter this problem if the default program is 'zsh -l'. If I make it to only 'zsh' I won't encounter it anymore. However, the problem won't happen if I set it to 'bash -l'.

**Reply by bew** [reply]
What is the value of `echo $TERM` ?

**Reply by sidzwans** [reply]
xterm-256color

It is set in my .zshrc.

**Reply by wez** [reply]
Just so I'm clear on what's happening, are you saying that when you exit neovim you expect the neovim display to go away and be replaced with your shell display?

If so, that sounds related to the alternate screen mode.  Usually, a "full-screen" terminal program like vim will switch to the alternate screen while it is running, then at the end, exit the alternate screen.

For the behavior to be different only when running zsh as a login shell there must be something in your shell startup that is the cause.

When troubleshooting this sort of difference, I'd suggest running `zsh -l -x` and `zsh -x` and comparing the difference in the output.

**Comment by sidzwans** [comment]
> Just so I'm clear on what's happening, are you saying that when you exit neovim you expect the neovim display to go away and be replaced with your shell display?

Yes. I expect the screen to clear and return to previous view/prompt.

As of now, I set the default_prog to `zsh` (without the login argument) as I achieve the result as described above.

The output of opening vim and close it straight away for `zsh -l -x` and `zsh -x` look almost the same using vimdiff if disregard the prompt. But I don't have any idea what they mean. Sorin is kind of from Prezto zsh theme that I use. 

For `zsh -x`:

> +_async_zle_watcher:1> setopt localoptions noshwordsplit                                                                                                I
+_async_zle_watcher:2> typeset -gA ASYNC_PTYS ASYNC_CALLBACKS
+_async_zle_watcher:3> local worker=prompt_sorin
+_async_zle_watcher:4> local callback=prompt_sorin_async_callback
+_async_zle_watcher:6> [[ -n '' ]]
+_async_zle_watcher:22> [[ -n prompt_sorin_async_callback ]]
+_async_zle_watcher:23> async_process_results prompt_sorin prompt_sorin_async_callback watcher
+async_process_results:1> setopt localoptions unset noshwordsplit noksharrays noposixidentifiers noposixstrings
+async_process_results:3> local worker=prompt_sorin
+async_process_results:4> local callback=prompt_sorin_async_callback
+async_process_results:5> local caller=watcher
+async_process_results:6> local -a items
+async_process_results:7> local null=$'\C-@' data
+async_process_results:8> integer -l len pos num_processed has_next
+async_process_results:10> typeset -gA ASYNC_PROCESS_BUFFER
+async_process_results:13> zpty -r -t prompt_sorin data
+async_process_results:14> ASYNC_PROCESS_BUFFER[$worker]+=$'\C-@prompt_sorin_async_git 0 \'\' 0.0073738098 \'\'\C-@'
+async_process_results:15> len=45
+async_process_results:16> pos=1
+async_process_results:19> ((  ! len  ))
+async_process_results:19> ((  pos > len  ))
+async_process_results:23> ((  pos <= len  ))
+async_process_results:26> items=( '' )
+async_process_results:29> ASYNC_PROCESS_BUFFER[$worker]=$'prompt_sorin_async_git 0 \'\' 0.0073738098 \'\'\C-@'
+async_process_results:31> len=44
+async_process_results:32> ((  len > 1  ))
+async_process_results:33> pos=44
+async_process_results:36> has_next=1
+async_process_results:37> ((  1 == 5  ))
+async_process_results:41> [[ -z '' ]]
+async_process_results:23> ((  pos <= len  ))
+async_process_results:26> items=( prompt_sorin_async_git 0 '' 0.0073738098 '' )
+async_process_results:29> ASYNC_PROCESS_BUFFER[$worker]=''
+async_process_results:31> len=0
+async_process_results:32> ((  len > 1  ))
+async_process_results:36> has_next=0
+async_process_results:37> ((  5 == 5  ))
+async_process_results:38> items+=( 0 )
+async_process_results:39> prompt_sorin_async_callback prompt_sorin_async_git 0 '' 0.0073738098 '' 0
+prompt_sorin_async_callback:1> case prompt_sorin_async_git (prompt_sorin_async_git)
+prompt_sorin_async_callback:4> IFS=: +prompt_sorin_async_callback:4> read _git_target _git_post_target
+prompt_sorin_async_callback:8> _git_target=+prompt_sorin_async_callback:8> coalesce
+coalesce:5> return 1
+prompt_sorin_async_callback:8> _git_target=''
+prompt_sorin_async_callback:10> [[ -z '' ]]
+prompt_sorin_async_callback:12> [[ -n '' ]]
+async_process_results:40> ((  num_processed++  ))
+async_process_results:23> ((  pos <= len  ))
+async_process_results:13> zpty -r -t prompt_sorin data
+async_process_results:52> ((  num_processed  ))
+async_process_results:52> return 0




For` zsh -l -x`:
> +> setopt localoptions noshwordsplit                                                                                                                    I
+> typeset -gA ASYNC_PTYS ASYNC_CALLBACKS
+> local worker=prompt_sorin
+> local callback=prompt_sorin_async_callback
+> [[ -n '' ]]
+> [[ -n prompt_sorin_async_callback ]]
+> async_process_results prompt_sorin prompt_sorin_async_callback watcher
+> setopt localoptions unset noshwordsplit noksharrays noposixidentifiers noposixstrings
+> local worker=prompt_sorin
+> local callback=prompt_sorin_async_callback
+> local caller=watcher
+> local -a items
+> local null=$'\C-@' data
+> integer -l len pos num_processed has_next
+> typeset -gA ASYNC_PROCESS_BUFFER
+> zpty -r -t prompt_sorin data
+> ASYNC_PROCESS_BUFFER[$worker]+=$'\C-@prompt_sorin_async_git 0 \'\' 0.0061781406 \'\'\C-@'
+> len=45
+> pos=1
+> ((  ! len  ))
+> ((  pos > len  ))
+> ((  pos <= len  ))
+> items=( '' )
+> ASYNC_PROCESS_BUFFER[$worker]=$'prompt_sorin_async_git 0 \'\' 0.0061781406 \'\'\C-@'
+> len=44
+> ((  len > 1  ))
+> pos=44
+> has_next=1
+> ((  1 == 5  ))
+> [[ -z '' ]]
+> ((  pos <= len  ))
+> items=( prompt_sorin_async_git 0 '' 0.0061781406 '' )
+> ASYNC_PROCESS_BUFFER[$worker]=''
+> len=0
+> ((  len > 1  ))
+> has_next=0
+> ((  5 == 5  ))
+> items+=( 0 )
+> prompt_sorin_async_callback prompt_sorin_async_git 0 '' 0.0061781406 '' 0
+> case prompt_sorin_async_git (prompt_sorin_async_git)
+> IFS=: +> read _git_target _git_post_target
+> _git_target=+> coalesce
+> return 1
+> _git_target=''
+> [[ -z '' ]]
+> [[ -n '' ]]
+> ((  num_processed++  ))
+> ((  pos <= len  ))
+> zpty -r -t prompt_sorin data
+> ((  num_processed  ))
+> return 0

Regards.
